### ISSUE TITLE
Question: Is denoiser.py using v-prediction instead of x-prediction?

### ISSUE THREAD

**Issue by ZezFuture** [LTH14/JiT#3]
Hi, thanks for open-sourcing this project!

I have a question about the training objective in `denoiser.py`.  
From the paper I was expecting an **x-prediction** objective, but the current implementation looks more like **v-prediction**.

The relevant code:

```python
def forward(self, x, labels):
    labels_dropped = self.drop_labels(labels) if self.training else labels

    t = self.sample_t(x.size(0), device=x.device).view(-1, *([1] * (x.ndim - 1)))
    e = torch.randn_like(x) * self.noise_scale

    z = t * x + (1 - t) * e
    v = (x - z) / (1 - t).clamp_min(self.t_eps)

    x_pred = self.net(z, t.flatten(), labels_dropped)
    v_pred = (x_pred - z) / (1 - t).clamp_min(self.t_eps)

    # l2 loss
    loss = (v - v_pred) ** 2
    loss = loss.mean(dim=(1, 2, 3)).mean()

    return loss

**Comment by x2x5** [comment]
> Hi, thanks for open-sourcing this project!
> 
> I have a question about the training objective in `denoiser.py`. From the paper I was expecting an **x-prediction** objective, but the current implementation looks more like **v-prediction**.
> 
> The relevant code:
> 
> def forward(self, x, labels):
>     labels_dropped = self.drop_labels(labels) if self.training else labels
> 
>     t = self.sample_t(x.size(0), device=x.device).view(-1, *([1] * (x.ndim - 1)))
>     e = torch.randn_like(x) * self.noise_scale
> 
>     z = t * x + (1 - t) * e
>     v = (x - z) / (1 - t).clamp_min(self.t_eps)
> 
>     x_pred = self.net(z, t.flatten(), labels_dropped)
>     v_pred = (x_pred - z) / (1 - t).clamp_min(self.t_eps)
> 
>     # l2 loss
>     loss = (v - v_pred) ** 2
>     loss = loss.mean(dim=(1, 2, 3)).mean()
> 
>     return loss

model output is x0, and loss is use v.  this is the best combination in the experiment:

<img width="492" height="279" alt="Image" src="https://github.com/user-attachments/assets/4a65c7cc-f9ff-4ba1-8697-a0dcd4b93333" />

**Comment by LTH14** [comment]
@x2x5 ’s understanding is correct -- the prediction space is x, the loss space is v.
